File: brain2/faculties/event_bus.py
class EventBus:
    """
    A lightweight Pub/Sub bus decoupling the C++ fuzzy state from Python execution.
    The Three Bridges (Topology, Teaching, Curiosity) all pass messages here rather
    than holding hard references to each other.
    """
    def __init__(self):
        self._subscribers = {}

    def subscribe(self, topic, callback):
        if topic not in self._subscribers:
            self._subscribers[topic] = []
        self._subscribers[topic].append(callback)

    def publish(self, topic, data=None):
        if topic in self._subscribers:
            for callback in self._subscribers[topic]:
                try:
                    callback(data)
                except Exception as e:
                    print(f"[EventBus] Error in subscriber for {topic}: {e}")

# Global instance for the whole brain
bus = EventBus()

File: brain3/middleware/java/src/brain3/middleware/EventBus.java
package brain3.middleware;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

/**
 * A lightweight Pub/Sub bus decoupling the native C++ components from the Java execution.
 * Bridges pass messages here rather than holding hard references to each other.
 */
public class EventBus {
    
    private final Map<String, List<Consumer<Object>>> subscribers;
    
    // Singleton pattern as in brain2 (global instance)
    private static final EventBus INSTANCE = new EventBus();
    
    private EventBus() {
        this.subscribers = new HashMap<>();
    }
    
    public static EventBus getInstance() {
        return INSTANCE;
    }
    
    public void subscribe(String topic, Consumer<Object> callback) {
        subscribers.computeIfAbsent(topic, k -> new ArrayList<>()).add(callback);
    }
    
    public void publish(String topic, Object data) {
        List<Consumer<Object>> topicSubscribers = subscribers.get(topic);
        if (topicSubscribers != null) {
            for (Consumer<Object> callback : topicSubscribers) {
                try {
                    callback.accept(data);
                } catch (Exception e) {
                    System.err.println("[EventBus] Error in subscriber for " + topic + ": " + e.getMessage());
                }
            }
        }
    }
}
